### PR TITLE
Initial local tensor support

### PR DESCRIPTION
**Pull request by ndryden**
This begins the new H2 tensors project by adding the local tensor structure. These tensors don't current _do_ much (doing stuff with them will come in a subsequent PR with operator support), but they should accomplish the most important things: having memory and indexing it correctly.

@benson31: Your usual thorough review is appreciated.
@tbennun & @fiedorowicz1: Mostly bringing you in to check over the [tensor API](https://github.com/LLNL/DiHydrogen/blob/tensor/include/h2/tensor/tensor_base.hpp). Let me know if there's anything critical missing.

Two notes:
1. The code still needs to be formatted.
2. Lazy allocation (and some other operations, like `contiguous`) will be implemented later.

### CODE REVIEW

File: .gitlab/spack/environments/corona/spack.yaml [hunk 1]
```diff
@@ -216,6 +216,11 @@ spack:
       externals:
       - spec: ninja@1.11.0
         prefix: /usr/tce/packages/ninja/ninja-1.11.0
+    openblas:
+      buildable: false
+      externals:
+      - spec: openblas@0.3.15
```
Comment (ndryden) [inline .gitlab/spack/environments/corona/spack.yaml:222]: Wow, that's an old OpenBLAS, but whatever works.

File: include/h2/tensor/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  TARGET H2Core COMPONENT CORE SCOPE INTERFACE
+  INSTALL_PREFIX "${H2_CURRENT_INSTALL_PREFIX}"
+  SOURCES
+  fixed_size_tuple.hpp
+  raw_buffer.hpp
+  strided_memory.hpp
+  tensor_types.hpp
+  tensor_base.hpp
+  tensor.hpp
+  )
```
Comment (benson31) [inline include/h2/tensor/CMakeLists.txt:22]: This is fine for now, but we may want this to be in `H2Tensor` or something.

File: include/h2/tensor/tensor_types.hpp [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+
+/**
+ * Integer type used for data indices.
+ */
+using DataIndexType = std::int64_t;
+
+/**
+ * Maximum number of dimensions a tensor may have.
+ */
+static constexpr NDimType MAX_TENSOR_DIMS = 8;
```
Comment (benson31) [inline include/h2/tensor/tensor_types.hpp:210]: cuDNN defines a maximum dimension; we could grab that here when we have cuDNN, if we care that much (the current max dim is 8, so happy coincidence).

**Review by benson31 — APPROVED** [review]
LGTM

**Pull request merged**